api/hasura/actions/upload_profile_avatar.ts

```diff
@@ -0,0 +1,96 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+import { v4 as uuidv4 } from 'uuid';
+
+import { gql } from '../../../api-lib/Gql';
+import { deleteImage, uploadImage } from '../../../api-lib/s3';
+import {
+  composeHasuraActionRequestBodyWithSession,
+  HasuraUserSessionVariables,
+  updateProfileAvatarInput,
+} from '../../../src/lib/zod';
+import { parseBase64Image, resizeAvatar } from '../images';
+
+export default async function handler(req: VercelRequest, res: VercelResponse) {
+  // parse the input
+  const {
+    input: { object: input },
+    session_variables: sessionVariables,
+  } = composeHasuraActionRequestBodyWithSession(
+    updateProfileAvatarInput,
+    HasuraUserSessionVariables
+  ).parse(req.body);
+
+  try {
+    const imageBytes = parseBase64Image(res, input.image_data_base64);
+    if (!imageBytes) {
+      return res;
+    }
+
+    // Figure out if there was a previous avatar, because we'll need to delete it
+    const { profiles_by_pk } = await gql.q('query')({
+      profiles_by_pk: [
+        {
+          id: sessionVariables.hasuraProfileId,
+        },
+        {
+          avatar: true,
+        },
+      ],
+    });
+
+    // see if there was a previous avatar so we can delete it after new one is saved
+    let previousAvatar: string | undefined = undefined;
+    if (profiles_by_pk) {
+      previousAvatar = profiles_by_pk.avatar;
+    }
+
+    // resize and crop the avatar
+    const avatarJpeg = await resizeAvatar(imageBytes);
+
+    // generate a filename for the new avatar
+    const fileName = uuidv4() + '.jpg';
+
+    // Uploading files to the bucket
+    try {
+      await uploadImage(fileName, avatarJpeg);
+    } catch (err: any) {
+      return res.status(500).json({
+        error: '500',
+        message: err.message
+          ? 'error uploading to s3: ' + err.message
+          : 'Unexpected error uploading file',
+      });
+    }
+
+    // save the new image id in the db
+    const mutationResult = await gql.q('mutation')({
+      update_profiles_by_pk: [
+        {
+          _set: { avatar: fileName },
+          pk_columns: { id: sessionVariables.hasuraProfileId },
+        },
+        {
+          id: true,
+          avatar: true,
+          address: true,
+        },
+      ],
+    });
+
+    if (mutationResult.update_profiles_by_pk) {
+      if (previousAvatar) {
```

Comment:
let's first check the `previousAvatar` existence here, and clean up the code I mentioned above

Reply:
I think it has to happen in the order as written, because previous has to be fetched before the mutation, otherwise it will be the mutated value

Reply:
Its possible to do an event trigger here, so the delete image can happen async in the background (as its not time sensitive nor critical). In the event handler just do a check on the payload for the old vs new column value for avatar column and execute the s3 delete.